### PR TITLE
[aws] Refactor ec2_group - also fixes #37160

### LINKED ISSUE #37160: ec2_group - 'purge_rules: yes' returns RulesPerSecurityGroupLimitExceeded when SG is maxed out
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
ec2_group

##### ANSIBLE VERSION
```
ansible 2.5.0rc1
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/elijah/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible-2.5.0rc1-py2.7.egg/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.6 (default, Nov 23 2017, 15:49:48) [GCC 4.8.4]

```

##### CONFIGURATION
`
DEFAULT_FORKS(/etc/ansible/ansible.cfg) = 8
DEFAULT_TIMEOUT(/etc/ansible/ansible.cfg) = 30
DEFAULT_VAULT_PASSWORD_FILE(env: ANSIBLE_VAULT_PASSWORD_FILE) = /home/elijah/.ansible/vault/ansicloud-cms-stage
DEPRECATION_WARNINGS(/etc/ansible/ansible.cfg) = False
PERSISTENT_CONNECT_TIMEOUT(/etc/ansible/ansible.cfg) = 30
`

##### SUMMARY
Using ec2_group module, I set `purge_rules: yes`, which should purge the non-existent rules. The existing group has 50 rules in it, which is our max. I updated the rules to contain only about 40 rules. Using ANSIBLE_DEBUG, I don't see the detail on if it is actually deleting the non-existent rules first or if it would try to do them last, which it appears to be doing it last. 

##### STEPS TO REPRODUCE
Working on this part.

##### EXPECTED RESULTS
The Security Group to be updated with the new rules. 

##### ACTUAL RESULTS
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ClientError: An error occurred (RulesPerSecurityGroupLimitExceeded) when calling the AuthorizeSecurityGroupIngress operation: The maximum number of rules per security group has been reached.
fatal: [Sg1]: FAILED! => {"attempts": 1, "changed": false, "error": {"code": "RulesPerSecurityGroupLimitExceeded", "message": "The maximum number of rules per security group has been reached."}, "msg": "Unable to authorize in for ip 69.31.33.0/24 security group 'Sg1' - An error occurred (RulesPerSecurityGroupLimitExceeded) when calling the AuthorizeSecurityGroupIngress operation: The maximum number of rules per security group has been reached.", "response_metadata": {"http_headers": {"connection": "close", "date": "Thu, 08 Mar 2018 01:33:46 GMT", "server": "AmazonEC2", "transfer-encoding": "chunked"}, "http_status_code": 400, "request_id": "a68c4348-6225-4a2a-929d-b0b9d3f8c683", "retry_attempts": 0}}
```

### PR DESCRIPTION
**Pull request by s-hertel**
Remove duplicate logic
Replace nested for loops with list comprehensions
Purge rules before adding new ones in case sg has maximum permitted rules

##### SUMMARY
Right now ec2_group is sprawling and hard to read. There is a lot of duplicate logic and due to its structure, when adding features more duplicate logic is usually necessary. It's doing odd things like deleting rules from a dictionary to determine which ones are left over to purge, whereas this could be calculated before making those calls. The module is also inefficiently calling authorize/revoke methods once per rule, while the boto3 methods accept a list, and the methods could be called just once for the whole module. Refactoring will also allow me to add diff mode to the module quite easily.

I also added tests for check mode, of which there were none.

Fixes #37160

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/ec2_group.py

##### ANSIBLE VERSION
```
2.6.0
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ec2_group.py [hunk 1]
```diff
@@ -706,48 +795,8 @@ def main():
     client = boto3_conn(module, conn_type='client', resource='ec2',
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_group.py]: This isn't needed anymore, you can do `module.client('ec2')` instead :smile:

File: lib/ansible/modules/cloud/amazon/ec2_group.py [hunk 2]
```diff
@@ -907,80 +852,68 @@ def main():
...
+
+        # Authorize new rules
+        changed = add_new_permissions(client, module, new_ingress_permissions, new_egress_permissions, group['GroupId'], changed)
+
+        # Update existing rules
+        changed = update_existing_rules(client, module, group['GroupId'], ingress=(present_ingress, current_ingress, 'in'),
+                                        egress=(present_egress, current_egress, 'out'), changed=changed)
 
     if group:
+        time.sleep(5)
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_group.py]: Using a sleep here because I made the module a lot faster and the updates may not be finished applying by the time it gets here. Looking at adding a few custom waiters instead of the sleep hack.

Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_group.py]: Definitely want to only do this sleep if changed is true, since we don't have to wait for propagation if no changes were made.

File: lib/ansible/modules/cloud/amazon/ec2_group.py [hunk 3]
```diff
@@ -592,27 +538,15 @@ def serialize_group_grant(group_id, rule):
...
+    permission = {'IpProtocol': rule['IpProtocol'],
+                  'FromPort': rule.get('FromPort'),
+                  'ToPort': rule.get('ToPort')}
     if 'GroupId' in grant:
-        permission = {'IpProtocol': rule['IpProtocol'],
-                      'FromPort': fromPort,
-                      'ToPort': toPort,
-                      'UserIdGroupPairs': [{'GroupId': grant['GroupId']}]
-                      }
+        permission.update(UserIdGroupPairs=[{'GroupId': grant['GroupId']}])
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_group.py]: Is there a reason you have `.update` here instead of `permission['UserIdGroupPairs'] = .....`?

**Review by ryansb — COMMENTED** [review]
Nice use of namedtuple here, couple minor requests.

File: lib/ansible/modules/cloud/amazon/ec2_group.py (hunk 3 above)
Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/ec2_group.py]: For absolutely no reason. Fixed those.

**Comment by ElijahLynn** [comment]
Thanks for the work on this. Do you know if this has a chance at the 2.5 release or will it be 2.6?

File: lib/ansible/modules/cloud/amazon/ec2_group.py [hunk 4]
```diff
@@ -645,21 +579,181 @@ def fix_port_and_protocol(permission):
...
+            group = client.create_security_group(**params)
+        except (BotoCoreError, ClientError) as e:
+            module.fail_json_aws(e, msg="Unable to create security group")
+        # When a group is created, an egress_rule ALLOW ALL
+        # to 0.0.0.0/0 is added automatically but it's not
+        # reflected in the object returned by the AWS API
+        # call. We re-read the group for getting an updated object
+        # amazon sometimes takes a couple seconds to update the security group so wait till it exists
+        while True:
+            group = get_security_groups_with_backoff(client, GroupIds=[group['GroupId']])['SecurityGroups'][0]
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_group.py]: I worry that this will *cause* throttling as this is going to call repeatedly with no backoff until throttling occurs.

Might be worth adding a short sleep before each retry at the very least.

File: lib/ansible/modules/cloud/amazon/ec2_group.py [hunk 5]
```diff
@@ -907,80 +852,69 @@ def main():
...
+        for rule_id in current_egress:
+            if rule_id not in user_egress_rule_ids:
+                if rules_egress:
+                    purge_egress_rule_ids.append(rule_id)
+                else:
+                    if not rule_id.endswith('0.0.0.0/0'):
+                        purge_egress_rule_ids.append(rule_id)
+
+        # Revoke old rules
+        for rules_to_purge, revoke_permissions, rule_type in [(purge_ingress_rule_ids, [], 'in'), (purge_egress_rule_ids, [], 'out')]:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_group.py]: With the excellent refactor being undertaken here, I wonder if the purging shouldn't be moved into a function

**Comment by s-hertel** [comment]
@ElijahLynn I think this is probably too invasive to be backported to 2.5.

**Comment by ElijahLynn** [comment]
@s-hertel Thanks!

**Comment by s-hertel** [comment]
Things still to do:
* add suboptions to the arg spec for rule validation
* wait for updates to complete instead of sleep() hack

**Comment by s-hertel** [comment]
Todo:
* Stabilize module for #36062

File: lib/ansible/modules/cloud/amazon/ec2_group.py [hunk 6]
```diff
@@ -288,28 +288,38 @@
...
-from ansible.module_utils.basic import AnsibleModule
-from ansible.module_utils.ec2 import boto3_conn
-from ansible.module_utils.ec2 import get_aws_connection_info
-from ansible.module_utils.ec2 import ec2_argument_spec
-from ansible.module_utils.ec2 import camel_dict_to_snake_dict
-from ansible.module_utils.ec2 import HAS_BOTO3
+from time import sleep
+from collections import namedtuple
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import boto3_conn, get_aws_connection_info, ec2_argument_spec, camel_dict_to_snake_dict
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_group.py:294]: Don't need to keep the `get_aws_connection_info` or `ec2_argument_spec` since AWSModule has `.client` now.

File: lib/ansible/modules/cloud/amazon/ec2_group.py [hunk 7]
```diff
@@ -840,147 +874,68 @@ def main():
...
+                    else:
+                        if not rule_id.endswith('0.0.0.0/0'):
+                            purge_egress_rule_ids.append(rule_id)
+
+        # Assemble ip permissions to remove
+        revoke_ingress, revoke_egress = assemble_permissions_to_revoke(purge_ingress_rule_ids, purge_egress_rule_ids,
+                                                                       current_ingress, current_egress)
+
+        # Revoke old rules
+        changed = remove_old_permissions(client, module, revoke_ingress, revoke_egress, group['GroupId'], changed)
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_group.py:927]: Instead of passing `changed` in on all these, can you make these assignments all `changed |= [method](.....)`?

File: lib/ansible/modules/cloud/amazon/ec2_group.py [hunk 8]
```diff
@@ -840,147 +874,68 @@ def main():
...
+        # Authorize new rules
+        changed = add_new_permissions(client, module, new_ingress_permissions, new_egress_permissions, group['GroupId'], changed)
+
+        # Update existing rules
+        changed = update_existing_rules(client, module, group['GroupId'], ingress=(present_ingress, current_ingress, 'in'),
+                                        egress=(present_egress, current_egress, 'out'), changed=changed)
 
     if group:
+        if changed:
+            sleep(5)
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_group.py:938]: Is this because we're missing a complex waiter that checks the security group rule lists against the parameters?

File: lib/ansible/modules/cloud/amazon/ec2_group.py [hunk 9]
```diff
@@ -840,147 +874,68 @@ def main():
...
+                if ipv6:
+                    named_tuple_rule_list.extend(add_cidr_ip_list(module, ipv6, "ipv6", rule, rule_type, group['GroupId']))
+
+        # List comprehensions for rules to add, rules to modify, and rule ids to determine purging
+        new_ingress_permissions = [r.ip_permission for r in named_tuple_ingress_list if r.rule_id not in current_ingress]
+        new_egress_permissions = [r.ip_permission for r in named_tuple_egress_list if r.rule_id not in current_egress]
+        present_ingress = [r for r in named_tuple_ingress_list if r.rule_id in current_ingress]
+        present_egress = [r for r in named_tuple_egress_list if r.rule_id in current_egress]
+        user_ingress_rule_ids = [r.rule_id for r in named_tuple_ingress_list]
+        user_egress_rule_ids = [r.rule_id for r in named_tuple_egress_list]
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_group.py:895]: :1st_place_medal:

**Review by ryansb — CHANGES_REQUESTED** [review]
